### DISCUSSION TITLE
Approval system based on branches; branch management from the interface

### DISCUSSION THREAD

**Discussion by ghost** [netbox-community/netbox#14931] [Ideas]
Hi,

I discovered the StagedChange and Branch classes
https://github.com/netbox-community/netbox/blob/3d941411d438f77b66d2036edf690c14b459af58/netbox/extras/models/staging.py#L21
and realized that this is not something you can use from the interface.

Most of the job is done, we could now:
* Display an interface for people to create new branches
* People will make their changes then ask for branch merge (I think the approval system should be hookable so we can create a module to change its behaviour. Some people might only want the administrator approval while, in our case, the approver will be all the tenants impacted by the changes, we might also want p out of n approvers, ... )
* Once the approval rules are validated, the branch is merged

I assume it was the initial goal of adding the branches. What do you think, would it be something we should implement in Netbox as a native feature?

There was a similar project with dolthub, even if all the sources mention Nautobot (Netbox fork), we can see on the first drafts that the integration was initially meant for netbox. Some links about the project:
* https://www.dolthub.com/blog/2021-09-24-announcing-nautobot-on-dolt/
* https://www.dolthub.com/blog/2023-05-10-nautobot-and-dolt/
By the way: they are working on a "postgres-flavored" dolthub, this could be interesting but since the branch system is here I don't see a reason for it now.

**Comment by jeremystretch** [comment]
> Most of the job is done

I'm afraid not. The low-level API that exists today is not suitable for session-based workflows as it leverages (abuses, maybe) database transactions to effect "branches." It's not feasible to simply graft a user interface on this as there's no feasible mechanism available to track state across multiple requests.

> There was a similar project with dolthub

A solution which requires users to lock themselves into a proprietary, experimental database with an uncertain future isn't much of a solution IMO. Call me a skeptic.

**Reply by ghost** [reply]
Hi @jeremystretch 
First thing, thank you for answering so fast.
Also, I want to make clear that I just mentioned dolthub the final UX they achieved is what I expect, but I am by no mean in favor of using it instead of postgres.


From what I saw, it stores the content of the transactions in a json inside of the branch instance ?

If the code `with checkout(...)` work, then I assumed we could:
1. keep only the name of the active branch in the session
2. use a middleware: if the session contains an active branch, then load this branch and use the `with checkout` around all the underlying execution.

Could you give me an insight of the current issues/limitations?
Is there an existing roadmap to improve the branch system?

**Comment by ypid-work** [comment]
I just found this thread when I searched for Dolt, a technology I find interesting but have not played much with. Anyway.

If people find this or are subscribed, https://github.com/netboxlabs/netbox-branching exists now and [has recently entered General Availability](https://netboxlabs.com/blog/netbox-branching-enters-general-availability/).
